frontend/src/metabase/dashboard/selectors.ts

```diff
@@ -439,12 +475,19 @@ export const getMissingRequiredParameters = createSelector(
  * It's a memoized version, it uses LRU cache per card identified by id
  */
 export const getQuestionByCard = createCachedSelector(
-  [(_state: State, props: { card: Card }) => props.card, getMetadata],
+  [
+    (_state: State, props: { card: Card | VirtualCard }) => props.card,
+    getMetadata,
+  ],
   (card, metadata) => {
     return isQuestionCard(card) ? new Question(card, metadata) : undefined;
   },
 )((_state, props) => {
-  return props.card.id;
+  // Sometime card doesn't have an id.
+  // In that case, we use the stringified version of the card as a cache key.
+  // That's nless than ideal, but it avoids flooding the console with warnings
+  // and it's still better than using `undefined` as a cache key.
+  return props.card.id ?? JSON.stringify(props.card).slice(0, 50);
```

Comment:
This may be not a reliable identifier for cards without IDs, for instance for heading cards we get the same instance via key `'{"name":null,"dataset_query":{},"display":"heading'`. I'll create a task in M5 to address this